### PR TITLE
Fixes for Secure Participant Discovery

### PR DESCRIPTION
**Pull request by kuznetsovmoci**

### CODE REVIEW

File: dds/DCPS/transport/rtps_udp/RtpsUdpReceiveStrategy.cpp [hunk 1]
```diff
@@ -653,7 +653,10 @@ bool RtpsUdpReceiveStrategy::decode_payload(ReceivedDataSample& sample,
   const DatawriterCryptoHandle writer_crypto_handle = link_->writer_crypto_handle(sample.header_.publication_id_);
   const CryptoTransform_var crypto = link_->security_config()->get_crypto_transform();
 
-  if (writer_crypto_handle == DDS::HANDLE_NIL || !crypto) {
+  const EndpointSecurityAttributesMask esa = link_->security_attributes(sample.header_.publication_id_);
+  bool payload_protected = esa & ENDPOINT_SECURITY_ATTRIBUTES_FLAG_IS_PAYLOAD_PROTECTED;
```
Comment (jrw972) [inline dds/DCPS/transport/rtps_udp/RtpsUdpReceiveStrategy.cpp]: const

File: dds/DCPS/RTPS/Sedp.h [hunk 2]
```diff
@@ -178,6 +178,9 @@ class Sedp : public DCPS::EndpointManager<ParticipantData_t> {
   };
 
   Spdp& spdp_;
+  DCPS::SequenceNumber seq_;
+  DCPS::SequenceNumber durable_seq_;
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Sedp.h]: Is this used anywhere?

File: dds/DCPS/RTPS/Sedp.cpp [hunk 3]
```diff
@@ -3753,12 +3753,14 @@ DDS::ReturnCode_t
 Sedp::write_dcps_participant_secure(const Security::SPDPdiscoveredParticipantData& msg,
                                     const RepoId& part)
 {
-  static DCPS::SequenceNumber sequence = 0;
-
   DCPS::RepoId remote_reader(part);
-  remote_reader.entityId = ENTITYID_SPDP_RELIABLE_BUILTIN_PARTICIPANT_SECURE_READER;
+  if (part != GUID_UNKNOWN) {
+    remote_reader.entityId = ENTITYID_SPDP_RELIABLE_BUILTIN_PARTICIPANT_SECURE_READER;
+    return dcps_participant_secure_writer_->write_dcps_participant_secure(msg, remote_reader, seq_++);
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Sedp.cpp]: Don't increment seq_ here

File: dds/DCPS/RTPS/Sedp.cpp [hunk 4]
```diff
@@ -3753,12 +3753,14 @@ DDS::ReturnCode_t
...
+  }
 
-  return dcps_participant_secure_writer_->write_dcps_participant_secure(msg, remote_reader, ++sequence);
+  DCPS::SequenceNumber sequence = DCPS::SequenceNumber::SEQUENCENUMBER_UNKNOWN();
+  return dcps_participant_secure_writer_->write_dcps_participant_secure(msg, remote_reader, sequence);
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Sedp.cpp]: Use seq_.  Also rename it so we know what it does (`participant_secure_sequence_`)

File: dds/DCPS/RTPS/Sedp.cpp [hunk 5]
```diff
@@ -2887,6 +2887,7 @@ Sedp::Writer::Writer(const RepoId& pub_id, Sedp& sedp)
   header_.guidPrefix[9] = pub_id.guidPrefix[9];
   header_.guidPrefix[10] = pub_id.guidPrefix[10];
   header_.guidPrefix[11] = pub_id.guidPrefix[11];
+  seq_ = seq_init;
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Sedp.cpp]: Use constructor member initializer list.

File: dds/DCPS/RTPS/Sedp.cpp [hunk 6]
```diff
@@ -3146,7 +3146,7 @@ Sedp::Writer::write_dcps_participant_secure(const Security::SPDPdiscoveredPartic
     return DDS::RETCODE_ERROR;
   }
 
-  return write_parameter_list(plist, reader, sequence);
+return write_parameter_list(plist, reader, sequence);
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Sedp.cpp]:
```suggestion
  return write_parameter_list(plist, reader, sequence);
```

File: dds/DCPS/transport/rtps_udp/RtpsUdpReceiveStrategy.cpp [hunk 7]
```diff
@@ -653,7 +653,10 @@ bool RtpsUdpReceiveStrategy::decode_payload(ReceivedDataSample& sample,
   const DatawriterCryptoHandle writer_crypto_handle = link_->writer_crypto_handle(sample.header_.publication_id_);
   const CryptoTransform_var crypto = link_->security_config()->get_crypto_transform();
 
-  if (writer_crypto_handle == DDS::HANDLE_NIL || !crypto) {
+  const EndpointSecurityAttributesMask esa = link_->security_attributes(sample.header_.publication_id_);
+  const bool payload_protected = esa & ENDPOINT_SECURITY_ATTRIBUTES_FLAG_IS_PAYLOAD_PROTECTED;
```
Comment (mitza-oci) [inline dds/DCPS/transport/rtps_udp/RtpsUdpReceiveStrategy.cpp]: FLAG_IS_VALID was checked in the old version.  Probably should be here too.

**Pull request merged**